protocols/kad/src/record.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Abstracts the Kademlia record store behaviour and provides default in-memory store
+
+use fnv::FnvHashMap;
+use multihash::Multihash;
+use std::borrow::Cow;
+
+/// The error record store may return
+#[derive(Clone, Debug, PartialEq)]
+pub enum RecordStorageError {
+    /// Store reached the capacity limit.
+    AtCapacity,
+    /// Value being put is larger than the limit.
+    ValueTooLarge,
+}
+
+/// The records that are kept in the dht.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Record {
+    /// Key of the record.
+    pub key: Multihash,
+    /// Value of the record.
+    pub value: Vec<u8>,
+}
+
+/// Trait for a record store.
+pub trait RecordStore {
```

Comment:
As mentioned, I think the methods here should be asynchronous. We should open an issue about that when the PR gets merged.